extensions/typescript-language-features/src/tsServer/server.ts

```diff
@@ -246,16 +250,19 @@ export class ProcessBasedTsServer extends Disposable implements ITypeScriptServe
 		}
 
 		this._requestQueue.enqueue(requestInfo);
-		this.sendNextRequests();
+		this.sendNextRequests(executeInfo.expectsResult && executeInfo.token);
 
 		return [result];
 	}
 
-	private sendNextRequests(): void {
+	private sendNextRequests(cancelToken?: vscode.CancellationToken | false): void {
 		while (this._pendingResponses.size === 0 && this._requestQueue.length > 0) {
 			const item = this._requestQueue.dequeue();
 			if (item) {
 				this.sendRequest(item);
+				if (cancelToken) {
```

Comment:
Can you put this logic in `this.tryCancelRequets` instead?

Reply:
It fits quite nicely there, but I'm not sure it will work: Cancellation.addData needs to add information to a request that will be posted to tsserver's webworker. If there is no pending request the code will have to wait until there is one.

For now I put in `peek` method on RequestQueue and a busy wait in tryCancelRequests, but it seems wrong. Could be better than what I had before, though. 

@jrieken can you provide advice here?

Reply:
The latest commit provides the current `request` to tryCancelRequest so that . However, it might actually be worse, since I believe that `request` will have been sent to tsserver already, and that `Cancellation.addData` won't actually transmit the request for cancellation. (again, @jrieken will know how this should work, I think)

I haven't worked out how to test this, so advice on that would be helpful as well.

Reply:
`Cancellation.addData` must be added before sending the request and the function that it returns need to be invoked in case of cancellation. I have some local changes on top of this PR that I am attempting to push. There was also an issue on the retrieving end. For that I for now inlines the implementation with a fix

Reply:
8639b4b3bd5b827dff405863bb4e01111c8deb4b it is. It would be nicer if this could be done with `serverProcess.browser.ts` but I couldn't find a neat/minimal way to doing that